### DISCUSSION TITLE
Why use MQTT rather than HTTP?

### DISCUSSION THREAD

**Discussion by wayner9** [arendst/Tasmota#20010] [Q&A]
It seems like the primary way that folks control Tasmota devices is MQTT, despite that fact that you can do pretty much everything with HTTP Get commands.  Why is this?  Does MQTT provide a better platform for smarthome devices?  One downside to MQTT is that you need a third party device (an MQTT) to sit in the middle between your device on your HA system or whatever is interfacing with the end Tasmota devices.

I have a Control4 Home Automation system. With that system you generally have drivers that control devices via HTTP, although MQTT is supported through a third-party driver as well.  But those folks using MQTT are more DIYers rather than the core Control4 model which is a heavily dealer supported model.

**Comment by sfromis** [comment]
A major reason for MQTT is that that you cannot "do pretty much everything" with http. MQTT is both ways, with Tasmota devices publishing availability and status info, also sensor values, automatically getting to subscribers. Another factor is that using http is slower (needs more network roundtrips), and lacks other features of MQTT, like retained messages enabling easy automatic discovery, or publishing to group topics.

As a reasonable smarthome setup will have some sort of backend platform anyway, this is a great place for also having the lightweight MQTT broker, thus no additional devices, and certainly not "third party".

**Reply by barbudor** [reply]
HTTP is one-to-one
MQTT support one-to-many: any service who needs to receive some info just subscribe to this information and receive it at no additional cost for the sender

HTTP is synchronous: you have to poll to get information. You even have to poll if you want to check that there is new information
MQTT is asynchronous: a device sends update when it has to. It doesn't need to be polled

Now the "folks" here are convinced and that the choice of Tasmota
You are free to think differently and build your own solution

**Reply by wayner9** [reply]
OK thanks folks.  Home Assistant is a very popular DIY HA system with Tasmota devices.  Does it have a built in MQTT broker?

I thought that the typical implementation is that a MQTT broker ran on your network that was separate from your HA controller.  It seems that is not the case.

**Reply by barbudor** [reply]
A broker is a separated service/daemon/process 
It can ran on the same "computer" or another
You can deploy it separately than the home automation system or it can be bundled.
Home Assistant can provide a MQTT broker as part of it's setup or it can use one sepately installed.

**Reply by wayner9** [reply]
Thanks.  I installed a MQTT broker in a Docker Container on my unRAID server which runs 24x7.  But I have not really used it, other than occasional playing around.

**Comment by tsmoneta** [comment]
I use HTTP in my environment; MQTT is disabled on Tasmota. The reason is that I started my home automation journey back around 1998, before MQTT or Home Assistant existed. I also have diverse devices that don't support MQTT. So I started building my home automation around HTTP.

To address the comment about having to poll to get information; that's incorrect. Tasmota can send its logging to a log server. I use rsyslog, which can parse the messages from Tasmota and invoke asynchronous action routines. They can propagate the messages as needed and/or update variables.

I also wrote my own web-based home automation panel which then uses the information from the rsyslog action routines, and sends HTTP messages to devices to control them.

![Screenshot 2023-11-14 13 22 10](https://github.com/arendst/Tasmota/assets/131212002/1141a080-0fc1-4830-9f37-b1542b9d0815)

This also lets me add smart controls to my security system (25 cameras), appropriate for each camera.

![image](https://github.com/arendst/Tasmota/assets/131212002/03758731-2184-4353-ad2b-2eaa33db228e)

Coupled with Wireguard (and Tailscale for backup), I can run with all ports closed, all devices bidirectionally prohibited from communicating to the Internet.

If I had switched to MQTT, I'd still need an HTTP infrastructure for the non-MQTT devices. There's always more than one way to accomplish a task.

**Reply by sfromis** [reply]
That way you are still losing significant features of Tasmota, this is not just about the polling. And using syslog to mimic some of what you get with MQTT is not a great workaround, as you do not get the speed and reliability.

**Comment by tsmoneta** [comment]
The speed is instant (subsecond). What features? Rsyslog gets the same status messages as MQTT, and HTTP has access to the full suite of Tasmota commands.

**Reply by sfromis** [reply]
Via MQTT you get things like availability updates, publishing to group topics, retained messages, easy automatic discovery, better deep sleep management, and multiple subscribers. Also authentication in the broker if you want that (I'm not using that). I'd really not like stepping down to only having syslog.

**Reply by tsmoneta** [reply]
A wget failure changes availability, and it's changed back when the module becomes reachable (e.g., incoming status). On the home automation page, a module status will change from on/off to ??? on loss of availability using that information.

The retained messages are provided by the processed rsyslog logging (the right column of my web-based home automation page). The server performs all processing. Anyone connecting to my home automation gets the same data.

I use all static addressing, so automatic discovery is a no-op.

None of my devices use deep sleep. I did experiment with it, and it worked fine (module is unavailable until it wakes). 

Multiple subscribers are to the MQTT broker; your web server provides the same function.

Not really losing function.


Like I said, I started doing this in 1998, so all these issues have been addressed, more than a decade before MQTT was broadly available.

**Reply by sfromis** [reply]
Well, if you do things like incessant polling, you can of course see no-response, but that is a workaround instead of actual availability when a device goes away.

You are not getting anything like retained messages via syslog handling. Remember that MQTT is bidirectional and multiple subscribers.

Static addressing does not give automatic discovery of device details.

No, a web server does not provide multiple subscribers, like one Tasmota device subscribing to messages from another. Sure, you can code something to do whatever, if that's your way of having fun.

Note: To be clear, I was not trying to talk about the many workarounds you may have created, but about it taking a lot of work (which you have done quite a bit of) to mimic what MQTT gives out of the box.

If your highly specialized setup works for you, great. It is just not a viable suggestion that others can easily replace MQTT by syslog.

**Comment by tsmoneta** [comment]
"Actual availability" is a timeout reported by the broker. A periodic poll will report loss of availability much faster. Recovered availability happens at the same time with MQTT or an rsyslog action routine, on contact.

What do you do with retained messages that event logging doesn't provide? Web servers have multiple concurrent clients too.

Coding in Home Assistant vs coding action routines for rsyslog is a wash in my book. I can write action routines in any language that Linux supports. And action routines can, of course, control multiple devices. When I come into a room, a proximity sensor can turn on multiple lights and devices, and reverse when I leave. The fan-out of actions is provided by the server and all actions occur (apparently) simultaneously as multiple background processes. 

I'm not trying to convince you to use HTTP. If you're happy with MQTT and a packaged home automation system, that's great. I have 104 Tasmota devices (currently) online, out of 146 total WiFi devices, and the prepackaged solutions don't work for me (or more importantly, my wife). In ten years, all this will probably end up being replaced by some AIs and robots anyway.

**Reply by sfromis** [reply]
Well, I'm not using any "packaged home automation systems" either, better not make assumptions. I'm doing things in my own programming language, some Node-Red, and also some cross-device automations via the Berry language.

One thing I do with retained messages is that another client can come online, and via a subscribe have a lot of info right away. No polling, and not database coding from me. Looks like you have coded something allowing you do do something similar. And really, web servers having concurrent clients has nothing to do with multiple subscribers getting messages, unless you have some MQTT-like coding sitting behind it.

Anyway, the topic was about reasons for using MQTT over HTTP. Not for comparing to your highly specialized setup where you have coded stuff to make things work. And of course not trying to make you do things differently, merely pointing out that other users have an easier path to a working setup by including MQTT instead of themselves coding MQTT-like features on top of syslog.

And I fully understand that you like to stick with your unusual setup, when you have done the work to make it work.

**Comment by mcrazvan** [comment]
> It seems like the primary way that folks control Tasmota devices is MQTT, despite that fact that you can do pretty much everything with HTTP Get commands. Why is this? Does MQTT provide a better platform for smarthome devices? One downside to MQTT is that you need a third party device (an MQTT) to sit in the middle between your device on your HA system or whatever is interfacing with the end Tasmota devices.
> 
> I have a Control4 Home Automation system. With that system you generally have drivers that control devices via HTTP, although MQTT is supported through a third-party driver as well. But those folks using MQTT are more DIYers rather than the core Control4 model which is a heavily dealer supported model.

Hi, can you help with some tips regarding Tasmota devices integration with Control4? Thanks
